File: ios/Classes/ScLoanFlutterPlugin.swift
//
//  ScLoan.swift
//  scgateway_flutter_plugin
//
//  Created by Ankit Deshmukh on 08/08/23.
//

import Foundation
import Flutter
import UIKit
import Loans

public class ScLoanFlutterPlugin: NSObject, FlutterPlugin {
    
    let currentViewController: UIViewController = (UIApplication.shared.delegate?.window??.rootViewController)!
      
    public static func register(with registrar: FlutterPluginRegistrar) {
      let channel = FlutterMethodChannel(name: "ScLoan_flutter_plugin", binaryMessenger: registrar.messenger())
      let instance = ScLoanFlutterPlugin()
      registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        guard let args = call.arguments as? Dictionary<String, Any> else { return }
        
        switch (call.method) {
            
        case "setup": do {
            guard let gateway = args["gateway"] as? String else {
                //SDK setup failure
                print("ScLoan setup failed: No gateway name provided")
                return
            }
            
            var env = SCLoanEnvironment.production
            
            if let hostEnv = args["env"] as? Int {
                env = SCLoanEnvironment(rawValue: hostEnv) ?? .production
            }
            
            ScLoan.instance.setup(
                config: ScLoanConfig(gatewayName: gateway, environment: env)) { success, error in
                    
                    if let err = error {
                        result(self.convertErrorToJsonString(error: err))
                    } else {
                        result(success?.data)
                    }
                }
        }
            
        case "apply": do {
            guard let interactionToken = args["interactionToken"] as? String else {
                //SDK setup failure
                print("ScLoan apply failed: No interactionToken provided")
                return
            }
            
            ScLoan.instance.apply(
                presentingController: currentViewController,
                loanInfo: ScLoanInfo(interactionToken: interactionToken)) { success, error in
                    
                    if let err = error {
                        result(self.convertErrorToJsonString(error: err))
                    } else {
                        result(success?.data ?? "success")
                    }
                }
        }
            
        case "pay": do {
            guard let interactionToken = args["interactionToken"] as? String else {
                //SDK setup failure
                print("ScLoan pay failed: No interactionToken provided")
                return
            }
            
            ScLoan.instance.pay(
                presentingController: currentViewController,
                loanInfo: ScLoanInfo(interactionToken: interactionToken)) { success, error in
                    
                    if let err = error {
                        result(self.convertErrorToJsonString(error: err))
                    } else {
                        result(success?.data)
                    }
                }
        }
            
        case "withdraw": do {
            guard let interactionToken = args["interactionToken"] as? String else {
                //SDK setup failure
                print("ScLoan withdraw failed: No interactionToken provided")
                return
            }
            
            ScLoan.instance.withdraw(
                presentingController: currentViewController,
                loanInfo: ScLoanInfo(interactionToken: interactionToken)) { success, error in
                    
                    if let err = error {
                        result(self.convertErrorToJsonString(error: err))
                    } else {
                        result(success?.data)
                    }
                }
        }

        case "service": do {
            guard let interactionToken = args["interactionToken"] as? String else {
                //SDK setup failure
                print("ScLoan service failed: No interactionToken provided")
                return
            }
            
            ScLoan.instance.service(
                presentingController: currentViewController,
                loanInfo: ScLoanInfo(interactionToken: interactionToken)) { success, error in
                    
                    if let err = error {
                        result(self.convertErrorToJsonString(error: err))
                    } else {
                        result(success?.data)
                    }
                }
        }
            
        default:
            result("Flutter method not implemented on iOS")
        }
    }
    
    private func convertErrorToJsonString(error: ScLoanError) -> String? {
        
        var errorDict : [String: Any?] = [
            "isSuccess": error.isSuccess,
            "errorCode": error.errorCode,
            "errorMessage": error.errorMessage,
        ]
        
        if let errorData = error.data {
            errorDict["data"] = errorData.toDictionary
        }
        
        return errorDict.toJsonString
    }
    
}

File: android/src/main/kotlin/com/example/scgateway_flutter_plugin/ScLoanFlutterPlugin.kt
package com.example.scgateway_flutter_plugin

import android.app.Activity
import androidx.appcompat.app.AppCompatActivity
import com.google.gson.Gson
import com.smallcase.loans.core.external.*
import com.smallcase.loans.core.internal.ScLoanEnvironment
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

class ScLoanFlutterPlugin(val getActivity: () -> Activity): MethodChannel.MethodCallHandler {

    private val TAG: String = "Android_ScLoan"

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {

        when (call.method) {
            "setup" -> {

                val env: Int? = call.argument("env")
                val gateway: String? = call.argument("gateway")

                val environment = when (env) {
                    0 -> ScLoanEnvironment.DEVELOPMENT
                    2 -> ScLoanEnvironment.STAGING
                    else -> ScLoanEnvironment.PRODUCTION
                }

                ScLoan.setup(
                    ScLoanConfig(environment = environment, gatewayName = gateway ?: "gatewaydemo"),
                    listener = object : ScLoanResult {
                        override fun onFailure(error: ScLoanError) {
                            result.error(error.toString(), null, null)
                        }

                        override fun onSuccess(response: ScLoanSuccess) {
                            result.success(response.toString())
                        }

                    }
                )
            }
            "apply" -> {
                val interactionToken: String? = call.argument("interactionToken")

                ScLoan.apply(
                    getActivity(),
                    ScLoanInfo(interactionToken ?: ""),
                    listener = object : ScLoanResult {
                        override fun onFailure(error: ScLoanError) {
                            result.error(error.toString(), null, null)
                        }

                        override fun onSuccess(response: ScLoanSuccess) {
                            result.success(response.toString())
                        }

                    }
                )
            }
            "pay" -> {
                val interactionToken: String? = call.argument("interactionToken")

                ScLoan.pay(
                    getActivity(),
                    ScLoanInfo(interactionToken ?: ""),
                    listener = object : ScLoanResult {
                        override fun onFailure(error: ScLoanError) {
                            result.error(error.toString(), null, null)
                        }

                        override fun onSuccess(response: ScLoanSuccess) {
                            result.success(response.toString())
                        }

                    }
                )
            }
            "withdraw" -> {
                val interactionToken: String? = call.argument("interactionToken")

                ScLoan.withdraw(
                    getActivity(),
                    ScLoanInfo(interactionToken ?: ""),
                    listener = object : ScLoanResult {
                        override fun onFailure(error: ScLoanError) {
                            result.error(error.toString(), null, null)
                        }

                        override fun onSuccess(response: ScLoanSuccess) {
                            result.success(response.toString())
                        }

                    }
                )
            }
            "service" -> {
                val interactionToken: String? = call.argument("interactionToken")

                ScLoan.service(
                    getActivity(),
                    ScLoanInfo(interactionToken ?: ""),
                    listener = object : ScLoanResult {
                        override fun onFailure(error: ScLoanError) {
                            result.error(error.toString(), null, null)
                        }

                        override fun onSuccess(response: ScLoanSuccess) {
                            result.success(response.toString())
                        }

                    }
                )
            }
        }
    }
}